scripts/lib/puppet_cache.py

```diff
@@ -0,0 +1,70 @@
+import hashlib
+import os
+import shutil
+import subprocess
+
+import yaml
+
+from .zulip_tools import parse_os_release, run, su_to_zulip
+
+ZULIP_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+PUPPET_MODULES_CACHE_PATH = os.path.join(ZULIP_PATH, "puppet_thirdparty")
+PUPPET_MODULES_HASH_PATH = os.path.join(PUPPET_MODULES_CACHE_PATH, "deps.sha256")
+PUPPET_DEPS_FILE_PATH = os.path.join(ZULIP_PATH, "puppet/deps.yaml")
+
+
+def generate_sha1sum_node_modules() -> str:
+    sha1sum = hashlib.sha1()
+    with open(PUPPET_DEPS_FILE_PATH, "rb") as fb:
+        sha1sum.update(fb.read().strip())
+    sha1sum.update(
```

Comment:
Hashing the concatenation of a bunch of strings isn’t *really* correct because you’ve lost information about where the concatenation points are. Consider hashing a JSON object instead.

(Yeah, the existing `generate_sha1sum_node_modules` has this bug too.)

Reply:
This was fixed in https://github.com/zulip/zulip/pull/18636.